File: electrical/src/main/java/resonantinduction/electrical/generator/TileMotor.java
package resonantinduction.electrical.generator;

import java.util.EnumSet;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.ForgeDirection;
import universalelectricity.api.energy.EnergyStorageHandler;
import calclavia.api.resonantinduction.IMechanicalNode;
import calclavia.lib.grid.INode;
import calclavia.lib.grid.INodeProvider;
import calclavia.lib.grid.NodeRegistry;
import calclavia.lib.prefab.tile.IRotatable;
import calclavia.lib.prefab.tile.TileElectrical;

/**
 * A kinetic energy to electrical energy converter.
 * 
 * @author Calclavia
 */
public class TileMotor extends TileElectrical implements IRotatable, INodeProvider
{
	protected IMechanicalNode node;

	/** Generator turns KE -> EE. Inverted one will turn EE -> KE. */
	public boolean isInversed = true;
	private byte gearRatio;

	public TileMotor()
	{
		energy = new EnergyStorageHandler(1000000);
		node = NodeRegistry.get(this, IMechanicalNode.class);

		if (node != null)
			node.setLoad(0.5);
	}

	public byte toggleGearRatio()
	{
		return gearRatio = (byte) ((gearRatio + 1) % 3);
	}

	@Override
	public void initiate()
	{
		super.initiate();
		node.reconstruct();
	}

	@Override
	public void invalidate()
	{
		node.deconstruct();
		super.invalidate();
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		if (node != null)
		{
			if (!isInversed)
			{
				receiveMechanical();
				produce();
			}
			else
			{
				produceMechanical();
			}
		}
	}

	public void receiveMechanical()
	{
		double power = node.getEnergy();

		long receive = energy.receiveEnergy((long) power, true);

		if (receive > 0)
		{
			double percentageUsed = receive / power;
			node.apply(-node.getTorque() * percentageUsed, -node.getAngularVelocity() * percentageUsed);
		}
	}

	public void produceMechanical()
	{
		long extract = energy.extractEnergy(energy.getEnergy(), false);

		if (extract > 0)
		{
			long torqueRatio = (long) ((gearRatio + 1) / 2.2d * (extract));

			if (torqueRatio > 0)
			{
				final double maxAngularVelocity = extract / (float) torqueRatio;

				final double maxTorque = (extract) / maxAngularVelocity;

				double setAngularVelocity = maxAngularVelocity;
				double setTorque = maxTorque;

				double currentTorque = Math.abs(node.getTorque());

				if (currentTorque != 0)
					setTorque = Math.min(setTorque, maxTorque) * (node.getTorque() / currentTorque);

				double currentVelo = Math.abs(node.getAngularVelocity());
				if (currentVelo != 0)
					setAngularVelocity = Math.min(+setAngularVelocity, maxAngularVelocity) * (node.getAngularVelocity() / currentVelo);

				node.apply(setTorque - node.getTorque(), setAngularVelocity - node.getAngularVelocity());
				energy.extractEnergy((long) Math.abs(setTorque * setAngularVelocity), true);
			}
		}

	}

	@Override
	public EnumSet<ForgeDirection> getInputDirections()
	{
		return getOutputDirections();
	}

	@Override
	public EnumSet<ForgeDirection> getOutputDirections()
	{
		EnumSet<ForgeDirection> dirs = EnumSet.allOf(ForgeDirection.class);
		dirs.remove(this.getDirection());
		dirs.remove(this.getDirection().getOpposite());
		return dirs;
	}

	@Override
	public ForgeDirection getDirection()
	{
		return ForgeDirection.getOrientation(this.getBlockMetadata());
	}

	@Override
	public void setDirection(ForgeDirection dir)
	{
		this.worldObj.setBlockMetadataWithNotify(xCoord, yCoord, zCoord, dir.ordinal(), 3);
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		isInversed = nbt.getBoolean("isInversed");
		gearRatio = nbt.getByte("gear");
	}

	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		nbt.setBoolean("isInversed", isInversed);
		nbt.setByte("gear", gearRatio);
	}

	@Override
	public <N extends INode> N getNode(Class<? super N> nodeType, ForgeDirection from)
	{
		if (from == getDirection() || from == getDirection().getOpposite())
		{
			if (nodeType.isAssignableFrom(node.getClass()))
				return (N) node;
		}

		return null;
	}
}

File: src/main/scala/resonantinduction/electrical/generator/TileMotor.scala
package resonantinduction.electrical.generator

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.material.Material
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{ChatComponentText, ResourceLocation}
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.common.util.ForgeDirection
import org.lwjgl.opengl.GL11
import resonant.api.tile.IRotatable
import resonant.lib.content.prefab.TElectric
import resonant.lib.grid.node.TSpatialNodeProvider
import resonant.lib.prefab.tile.TileAdvanced
import resonant.lib.render.RenderUtility
import resonant.lib.transform.vector.Vector3
import resonantinduction.core.Reference
import resonantinduction.mechanical.mech.grid.NodeMechanical

/**
 * A kinetic energy to electrical energy converter.
 *
 * @author Calclavia
 */
object TileMotor
{
  @SideOnly(Side.CLIENT)
  val model = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "motor.tcn"))
  @SideOnly(Side.CLIENT)
  val texture = new ResourceLocation(Reference.domain, Reference.modelPath + "motor.png")
}

class TileMotor extends TileAdvanced(Material.iron) with TElectric with TSpatialNodeProvider with IRotatable
{
  var mechNode = new NodeMechanical(this)
  {
    override def canConnect(from: ForgeDirection): Boolean =
    {
      connectionMask = 1 << getDirection.getOpposite.ordinal
      return super.canConnect(from)
    }
  }

  private var gearRatio = 0

  textureName = "material_wood_surface"
  normalRender = false
  isOpaqueCube = false
  nodes.add(dcNode)
  nodes.add(mechNode)

  def toggleGearRatio() = (gearRatio + 1) % 3

  /**
   * Called when the block is placed by a living entity
   * @param entityLiving - entity who placed the block
   * @param itemStack - ItemStack the entity used to place the block
   */
  override def onPlaced(entityLiving: EntityLivingBase, itemStack: ItemStack)
  {
    super.onPlaced(entityLiving, itemStack)

    if (!world.isRemote)
      mechNode.reconstruct()
  }

  override def update()
  {
    //TODO: Debug with free energy
    val deltaPower = 10000d //Math.abs(mechNode.power - dcNode.power)

    if (false && mechNode.power > dcNode.power)
    {
      //Produce electricity
      dcNode.buffer(deltaPower)
      //TODO: Resist mech energy
    }
    //    else if (dcNode.power > mechNode.power)
    else
    {
      //Produce mechanical energy
      val mechRatio = Math.pow(gearRatio + 1, 3) * 400

      if (mechRatio > 0)
      {
        mechNode.rotate(deltaPower, deltaPower / mechRatio)
        //TODO: Resist DC energy
      }
    }
  }

  @SideOnly(Side.CLIENT)
  override def renderDynamic(pos: Vector3, frame: Float, pass: Int): Unit =
  {
    GL11.glPushMatrix()
    GL11.glTranslatef(pos.x.toFloat + 0.5f, pos.y.toFloat + 0.5f, pos.z.toFloat + 0.5f)
    GL11.glRotatef(90, 0, 1, 0)
    RenderUtility.rotateBlockBasedOnDirection(getDirection)
    RenderUtility.bind(TileMotor.texture)
    TileMotor.model.renderAll()
    GL11.glPopMatrix()
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    gearRatio = nbt.getByte("gear")
  }

  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)
    nbt.setByte("gear", gearRatio.toByte)
  }

  override def toString: String = "[TileMotor]" + x + "x " + y + "y " + z + "z "

  override protected def configure(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (!world.isRemote)
    {
      gearRatio = (gearRatio + 1) % 3
      player.addChatComponentMessage(new ChatComponentText("Toggled gear ratio: " + gearRatio))
    }

    return true
  }
}